### PR TITLE
888 - Remove unaccepted snapshots from disk

### LINKED ISSUE #888: Remove incorrect snapshots from disk after redownload

### PR DESCRIPTION
**Pull request by kpudlik**
Resolved #888

### CODE REVIEW

File: src/test/scala/org/constellation/domain/redownload/RedownloadServiceTest.scala [hunk 1]
```diff
@@ -364,101 +333,45 @@ class RedownloadServiceTest
...
-        10L -> "e"
-      )
+      val acceptedSnapshots: SnapshotsAtHeight = Map(2L -> "a", 4L -> "b", 6L -> "c", 8L -> "d", 10L -> "e")
 
-      "returns snapshots to leave untouched" in {
-        val diff = redownloadService.calculateRedownloadPlan(acceptedSnapshots, majorityState)
-        diff.toLeave shouldEqual Map(
-          2L -> "a"
-        )
+      "ignores snapshots below ignore point" in {
```
Comment (marcinwadon) [inline src/test/scala/org/constellation/domain/redownload/RedownloadServiceTest.scala]: below ignore point (inclusive)

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 2]
```diff
@@ -123,6 +125,13 @@ class RedownloadService[F[_]](
     makeAttempt((stopAt + 1) % poolArray.length)
   }
 
+//  private def calculateRecent[K, V](data: Map[K, V], recent: Int)(implicit O: Ordering[K]): Map[K, V] = {
```
[the remark this answers is not in the document]
Comment (kpudlik) [reply src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: I think so. Now we persist as it is and then eventually trim/remove during redownload round.

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 3]
```diff
@@ -214,39 +218,79 @@ class RedownloadService[F[_]](
...
-      _ <- snapshotService
-        .syncBufferPull()
-        .flatMap(_.values.toList.traverse(checkpointAcceptanceService.accept(_)))
-        .attemptT
+      _ <- acceptCheckpointBlocks()
+
+      _ <- EitherT.liftF(logger.debug("RedownloadPlan has been applied succesfully."))
+    } yield ()
+
+  private[redownload] def getIgnorePoint(maxHeight: Long): Long = maxHeight - meaningfulSnapshotsCount * 2
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: It should stay `maxHeight - meaningfulSnapshotsCount`

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 4]
```diff
@@ -214,39 +218,79 @@ class RedownloadService[F[_]](
...
+      _ <- updateAcceptedSnapshots(plan)
+
+      _ <- EitherT.liftF(logger.debug("Removing unaccepted snapshots from disk."))
+      _ <- removeUnacceptedSnapshotsFromDisk()
 
       _ <- EitherT.liftF(logger.debug("Updating highest SnapshotInfo in SnapshotService."))
-      _ <- snapshotService.setSnapshot(highestSnapshotInfo).attemptT
+      _ <- updateHighestSnapshotInfo()
 
       _ <- EitherT.liftF(logger.debug("Accepting all the checkpoint blocks from the highest SnapshotInfo."))
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: `Accepting all the checkpoint blocks received during the redownload`

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 5]
```diff
@@ -55,21 +56,20 @@ class RedownloadService[F[_]](
...
-  private[redownload] val snapshotHeightRedownloadDelayInterval: Int =
-    ConfigUtil.constellation.getInt("snapshot.snapshotHeightRedownloadDelayInterval")
-
   private val logger = Slf4jLogger.getLogger[F]
 
   def persistCreatedSnapshot(height: Long, hash: String): F[Unit] =
     createdSnapshots.modify { m =>
       val updated = if (m.contains(height)) m else m.updated(height, hash)
-      (updated, ())
+      val limited = takeNHighest(updated, meaningfulSnapshotsCount + redownloadInterval)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: `meaningfulSnapshotsCount + 2 * redownloadInterval`

File: src/main/scala/org/constellation/domain/redownload/RedownloadService.scala [hunk 6]
```diff
@@ -55,21 +56,20 @@ class RedownloadService[F[_]](
...
+      (limited, ())
     }
 
   def persistAcceptedSnapshot(height: Long, hash: String): F[Unit] =
     acceptedSnapshots.modify { m =>
       val updated = m.updated(height, hash)
-      (updated, ())
+      val limited = takeNHighest(updated, meaningfulSnapshotsCount + redownloadInterval)
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/domain/redownload/RedownloadService.scala]: `meaningfulSnapshotsCount + 2 * redownloadInterval`

**Pull request merged**